### PR TITLE
Annotate `engine/core.lua`

### PR DESCRIPTION
**Pull request by Garanas**
Annotates the majority of functions in `engine/core.lua`

### CODE REVIEW

**Comment by Hdt80bro** [comment]
For future reference, the functions
```AudioSetLanguage, EntityCategoryContains, EntityCategoryFilterDown, GetBlueprint, GetDepositsAroundPoint, GetFocusArmy, GetGameTimeSeconds, GetUnitById, HasLocalizedVO, IsAlly, IsEnemy, IsNeutral, ParseEntityCategory, Random, SessionIsReplay, SetFocusArmy```
should be merged from `User.lua` and `Sim.lua` into `Core.lua`
and the functions `GetTimeForProfile, LuaDumpBinary, _ALERT, _TRACEBACK` should be created in `Core.lua`

**Comment by Hdt80bro** [comment]
For reference work, where we don't have code that gets read and worked on together, I'd recommend sorting it alphabetically instead of grouping it by usage. Just so finding stuff is easier

File: engine/Core.lua [hunk 1]
```diff
@@ -2,377 +2,436 @@
...
 ---@field [3] number
 
----
---  Create a target object
-function AITarget()
+---@class Rectangle     # A point-to-point based rectangle, where the first point is usually in the top left corner
+---@field x0 number
+---@field y0 number
+---@field x1 number
+---@field y1 number
```
[the remark this answers is not in the document]
Comment (Garanas) [reply engine/Core.lua:29]: On the sim side they are not, see also: https://github.com/FAForever/fa/blob/deploy/fafdevelop/lua/ScenarioFramework.lua#L1200

File: engine/Core.lua [hunk 2]
```diff
@@ -2,377 +2,436 @@
...
 
----
---  create an empty prefetch set
-function CreatePrefetchSet()
+-- # Quaternion / orientation related functions
+
+--- Converts euler angles to a quaternion
+---@param roll number float
+---@param pitch number float
+---@param yaw number float
```
[the remark this answers is not in the document]
Comment (Garanas) [reply engine/Core.lua:65]: I bluntly copied this from what was already there - I'll remove the float bits.

File: engine/Core.lua (hunk 1 above)
[the remark this answers is not in the document]
Comment (Hdt80bro) [reply engine/Core.lua:24]: On the sim-side... oh that's annoying.
So we need separate `Rectangle` types for sim and user? The annotation system really won't like that.
What if we added an `UserRect` alias then?

File: engine/Core.lua [hunk 3]
```diff
@@ -2,377 +2,436 @@
...
+--- Retrieves the game version, as set by `version.lua`
+---@return string
+function GetVersion()
 end
 
----
---  Difference of two vectors
-function VDiff()
+--- Create an empty prefetch set
+---@unknown
```
[the remark this answers is not in the document]
Comment (Hdt80bro) [reply engine/Core.lua:352]: The return type of `CreatePrefetchSet` is the actual `userdata: CPrefetchSet` C Object. Oddly enough, it appears to be the only function that doesn't wrap the C Object inside a table like `{_c_object: userdata}`.

**Comment by Hdt80bro** [comment]
How are you looking for stuff? If I'm looking for, say, `Vector`, 'V' is at the end of the alphabet, so I scroll to the bottom keeping track of the letters until 'V' is reached. Putting things out of alphabetical order doesn't allow me to do that, so I spend lots of time scrolling up and down looking for `Vector` until I give up and `ctrl+f` the thing, moving through all of the strings containing `Vector` until I find what I want.
I don't think grouping functions by relation is useful, since everything is a stub and no code is being actively worked on. Being purely for reference means that the only reason someone will come to this file is if they're looking for a specific function, so we should do what we can to help them.

**Comment by Garanas** [comment]
I usually always use `ctrl + f`, but I'll put it back in alphabetical order. I'm fine with that too.

**Comment by Garanas** [comment]
It is not entirely done - but it is as good as I can make it for now.

**Pull request merged**